src/utils/validations/component/validations/with-validation.hoc.js

```diff
@@ -0,0 +1,135 @@
+
+const withValidation = (WrappedComponent) => {
+  class WithValidation extends React.Component {
+    static contextType = ValidationsContext;
+
+    static propTypes = {
+      children: PropTypes.node, // Children elements
+      name: PropTypes.string, // Name to uniquely identify the component
+      value: PropTypes.string, // The current value of the component
+      error: PropTypes.oneOfType([ // The error validations that should be run against the value
+        PropTypes.func,
+        PropTypes.arrayOf(PropTypes.func)
+      ]),
+      validationTypes: PropTypes.arrayOf(PropTypes.string), // the types of validations to be run on the component
+      warning: PropTypes.oneOfType([ // The warnings validations that should be run against the value
+        PropTypes.func,
+        PropTypes.arrayOf(PropTypes.func)
+      ]),
+      info: PropTypes.oneOfType([ // The info validations that should be run against the value
+        PropTypes.func,
+        PropTypes.arrayOf(PropTypes.func)
+      ]),
+      onBlur: PropTypes.func // Custom function to be called when the component blurs
+    };
+
+    static defaultProps = {
+      error: [],
+      warning: [],
+      info: []
+    }
+
+    static contextTypes = {
+      addInput: PropTypes.func, // Function to allow child components to register with parent
+      removeInput: PropTypes.func, // Function to allow child components to unregister with parent
+      adjustCount: PropTypes.func // Function to allow child components to increment the count when a validation fails
+    }
+
+    state = {
+      hasError: false,
+      hasWarning: false,
+      hasInfo: false
+    };
+
+    componentDidMount = () => {
+      if (this.context && this.context.addInput && this.props.error.length) {
+        this.context.addInput(this.props.name, this.validate);
+      }
+    }
+
+    componentWillUnmount = () => {
+      if (this.context && this.context.removeInput && this.props.error.length) {
+        this.context.removeInput(this.props.name);
+      }
+    }
+
+    validate = async (types = this.props.validationTypes) => {
+      let result = true;
+      if (types.includes('error')) result = await this.runValidation('error');
+      if (result && types.includes('warning')) result = await this.runValidation('warning');
+      if (result && types.includes('info')) result = await this.runValidation('info');
+      return result;
+    }
+
+    updateValidationStatus = (type, errorStatus) => {
+      if (this.props.validationTypes.includes(type)) {
+        const { adjustCount } = this.context;
+        const stateProp = `has${type.charAt(0).toUpperCase() + type.slice(1)}`;
+        let newState = { [stateProp]: false };
+
+        if (errorStatus) {
+          adjustCount(type, errorStatus);
+          newState = { [stateProp]: errorStatus };
+        } else if (this.state[stateProp]) {
+          adjustCount(type);
+        }
+        this.setState(newState);
+      }
+      return !errorStatus;
```

Comment:
do we need it to return something?

Reply:
The return value is the boolean I pass into the resolve of the Promise, I could remove it and just pass it  a boolean instead I guess but not sure it makes much difference 🤔

Reply:
`errorStatus` is passed as an argument in the `updateValidationStatus` method, so we are returning that argument converted to boolean and negated. This method could just update the state without returning anything.

Reply:
Yeah on second thought it is more readable to go with your approach, I will change it and push it up later